Search in Doc Explorer

asiandrummer opened the pull request:
Search functionality in Documentation Explorer.
It's pretty basic with searching/getting results/clicking links - I'll incrementally add in more details in upcoming PRs.

EDIT: added no-search-result case and minimum character length

asiandrummer commented on src/components/DocExplorer.js:
```diff
@@ -45,8 +45,8 @@ export class DocExplorer extends React.Component {
     var navStack = this.state.navStack;
-    var isCurrentlyShown =
-      navStack.length > 0 && navStack[navStack.length - 1] === typeOrField;
+    var isCurrentlyShown = navStack.length > 0 &&
```
This is an unnecessary change - reverting.

asiandrummer commented:
<img width="417" alt="screen shot 2016-01-15 at 4 17 58 pm" src="https://cloud.githubusercontent.com/assets/1537569/12368803/e129101c-bba3-11e5-828b-d10a7660dc50.png">
<img width="392" alt="screen shot 2016-01-15 at 4 17 50 pm" src="https://cloud.githubusercontent.com/assets/1537569/12368804/e12c676c-bba3-11e5-832f-1637218e8fc7.png">

asiandrummer commented:
<img width="358" alt="screen shot 2016-01-15 at 4 43 39 pm" src="https://cloud.githubusercontent.com/assets/1537569/12369083/431f71be-bba7-11e5-824b-82f594bab7b9.png">
<img width="357" alt="screen shot 2016-01-15 at 4 39 30 pm" src="https://cloud.githubusercontent.com/assets/1537569/12369084/4320a566-bba7-11e5-9ae9-f2781e2e25ba.png">

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -56,7 +56,10 @@ export class DocExplorer extends React.Component {
-    this.state = { navStack: [] };
+    this.state = {
+      navStack: [],
+      searchValue: null
```
I think searchValue should probably be part of the nav stack rather than it's own thing, so that you can search, click on something, click back and be where you were before.

This could be done in a separate PR though.

asiandrummer replied:
Exactly; I would like the search action to be a part of navStack - something like:
navStack = [
  testType: { ... },
  graphiql_search: { value: ... },
];

leebyron replied:
A related note about navStack for the future: it would be cool if scroll position were also maintained as part of the back stack. Today if you click from a long scrolling view to another long scrolling view your scroll position is maintained when it should probably be reset to "0", but then when going "back" the original scroll position should be restored. It's a fairly small detail, but it will help make navigating larger schema easier on the mind.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -103,12 +114,25 @@ export class DocExplorer extends React.Component {
+          onClickField={this._onClickTypeOrField}
+        /> :
+        <span className="doc-alert-text">
+          Minimum search length is 3 characters.
```
I don't think we need to explicitly show this on the screen. Also, in a separate PR later we may want to remove this restriction so you could search for things like "id" but perhaps just change the search algorithm so it's not overwhelmed by < 3 ch searches.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -141,6 +165,15 @@ export class DocExplorer extends React.Component {
+              <input className="search-box-input"
+                onKeyUp={event => this._onSearch(event)}
+                type="text"
+                name="SearchBox"
```
I don't think you need `name=` here

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
+    const typeMap = schema.getTypeMap();
+
+    // Compose a search index for all possible arguments from every types.
+    let possibleArguments = [];
```
This isn't used anywhere below, maybe we should remove it for this PR.

Also, we don't want to compute this index on every render since it's likely to be expensive for larger schema. We only want to compute an index once per schema.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
+        matchedOn.push('Type Name');
+      }
+
+      if (type.description && type.description.indexOf(searchValue) !== -1) {
```
I would remove this for this initial PR - I'm concerned that searching over descriptions will be expensive for larger schema, and perhaps we'll want to do a specialized index for this. Let's come back to it in a separate PR once the main search functionality is working.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
+            );
+          }
+
+          if (field.args && field.args.length) {
```
probably want `else if` here so the same field doesnt appear twice if an argument also matches

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
+      }
+    });
+
+    if (matchedTypes.length === 0 &&
```
style nit:

```js
if (matchedTypes.length === 0 &&
    matchedFields.length === 0 &&
    matchedArgs.length === 0) {
  return ...
}

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +181,142 @@ export class DocExplorer extends React.Component {
+            let matches = field.args.filter(arg => {
+              return arg.name.indexOf(searchValue) !== -1;
+            });
+            Array.prototype.push.apply(matchedArgs, matches.map(arg =>
```
Since you're rendering fields here, you might want to add them to `matchedFields`. Also, rather than rendering the same field multiple times for each matched arg, you should just render all the matched args within this one field, moving the `matches.map` inside the `( )` selection below.

leebyron commented:
This is awesome! There are some comments inline, but this is pretty close to landing as an initial version of search.

asiandrummer commented:
Responded to comments:
- removed ch restriction for now
- got rid of the argument indexing, turns out we didn't need it after all
- removed specific match types and included wholesome within `search results`
- a field with multiple argument matches

asiandrummer commented:
<img width="362" alt="screen shot 2016-01-21 at 2 56 40 pm" src="https://cloud.githubusercontent.com/assets/1537569/12497273/a8a9cf78-c04f-11e5-98b3-f48137ff467a.png">

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -103,12 +114,21 @@ export class DocExplorer extends React.Component {
         />;
+    } else if (this.state.searchValue) {
+      title = 'Search Results';
+      content = <SearchDoc
```
nit: use the same styling here as in the else block below, line break and indent.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
+class SearchDoc extends React.Component {
+  shouldComponentUpdate(nextProps) {
+    return (this.props.schema !== nextProps.schema) ||
+      (this.props.searchValue !== nextProps.searchValue);
```
nit: no need for the wrapping () on each of these terms.

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
+
+    let matchedTypes = [];
+    let matchedFields = [];
+    let matchedArgs = [];
```
this is no longer being used, rm

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
+                <span>
+                  {matches.map(arg =>
+                    <span className="arg" key={arg.name}>
+                        <span className="arg-name">{arg.name}</span>
```
nit: indented 2sp too far

leebyron commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
+            );
+          } else if (field.args && field.args.length) {
+            let matches = field.args.filter(arg => {
+              return arg.name.indexOf(searchValue) !== -1;
```
you can make this a directly returned expression instead of a function block

```
const matches = field.args.filter(arg => arg.name.indexOf(searchValue) !== -1);
```

leebyron commented:
This looks great. I have a few other comments in there. One overarching minor thing: it's nice to use `const` instead of `let` as the default for defining variables, and use `let` as the exception when you need to reassign it in the future.

Once you've covered all the comments and updated it, merge away!

asiandrummer commented on src/components/DocExplorer.js:
```diff
@@ -148,6 +176,111 @@ export class DocExplorer extends React.Component {
+            let matches = field.args.filter(arg => {
+              return arg.name.indexOf(searchValue) !== -1;
+            });
+            matchedFields.push(
```
This will allow search results to match arguments/render fields every time. Fixed:
```
if (matches.length > 0) {
```

asiandrummer commented:
Thanks @leebyron!
Next diffs will include navStack integration and a newer format in Doc Explorer.
